### ISSUE TITLE
“Dithered” fill for the shape drawing routines?

### ISSUE THREAD

**Issue by Viza74** [nesbox/tic.computer#133]
With the cls command going above the basic 16 palette index we can achieve some kind of “dithered fill” for the screen, but the same trick is not working for the shape drawing commands (circ, rect, tri)
It could open up a couple of possibilities if the same “trick” could be used with those commands too.
(But I guess the “dither” in cls is simply caused by writing full bytes instead of half-bytes to the screen ram, so it is not usable in shapes, because it is not guaranteed that the are on byte borders)

**Comment by hseiken** [comment]
Dither patterns can be done (and usually are) by code, not the system, even in the old days.  

Dithers work in TIC80 already.

**Comment by Viza74** [comment]
@hseiken How are dithers working in TIC80?
You mean by poke-ing/pix-ing the colors?

Anyway, the essence of my suggestion is that TIC could offer "hardware" dither for fills - it is a fantasy console after all, not trying to emulate any real hardware (well maybe pico-8 if that can be called "real" or "hardware" :) ) 
So why not? Maybe it could lead to some interesting new-retro uses, and could be one of the distinctive features of TIC.

Also the cls command already shows some "dither-like" features, that was the inspiration for the suggestion. I know (guess) that is probably from just a side effect of writing bytes which are than interpreted as half-bytes, and understand that proper dithered fill routines could be much more work to implement, and could be a feature that few users use actively, so thats why this is just a suggestion. :)

**Comment by hseiken** [comment]
Trust me, I'd love to have automatic dithering too, but I think if we're to look at this as a learning platform too (pico8 is positioned as such as well), not just for games, but for US making them.  So yes, I long for dithering patterns being built in, but I think having direct control over them using poke segments and such allows more control and in the end, we can implement it differently every time for the occasion.  

So...I"m not against it, but I think if I'm going to become a better coder myself, I have to look at what can the machine do already and does it need to do it automatigically or should I get off my lazy ass and learn to really code it?  

This is one where...it could go either way.  I think it's a great suggestion, but as I suggested, maybe it needs to be thought of in terms of what TIC-80 is trying to accomplish as a project rather than what features it supports...because the features should reflect the project's goals.  

You are right, though, it IS a made up machine and could easily offer dither options, that's very true and from what I've seen of people's code to do dithering, is really only a couple of lines of code...but again, it's going to force everyone to handle dithering specifically in one way, the way the machine wants it to be handled.  

TL:DR:

It can be done, but the question is what is the goal of TIC, and does implementing it further that goal, or is it just a side quest?  Maybe the best thing is to wait for TIC-80 to have all of it's current menues and options that aren't even finished yet to be completed then revisit the question.

**Comment by Crutiatix** [comment]
I was working excatly on this last days. At first, I wanted to suggest to make from a color argument in drawing functions a optional function callback with reference arguments of position for every drawn pixel which would provide a handy way to custom dithers. A main reason is performance. Dithering generated pixel by pixel in lua code is veeeery slow. But then I figured out that even if the color callback in a drawing functions would by possible then an every pixel color would must to be decided by lua code, so performance would be not very good. Then I started to think simple. And created own drawing functions with sipmple chessboard dither using lines for drawing which is 10 times faster than pixel by pixel method. Here are codes for a rectangle and circle. An ellipse is big challenge for me and still working on it.

**Rect** - speed cca 0.35ms for size 240x136
```lua
function drect (x,y,w,h,c1,c2)
	if c2 then rect(x,y,w,h,c2) end
	local short = w < h and w or h
	local x1, x2, y1, y2
	for x1=0, w-1, 2 do
		y1 = 0
		x2 = w-x1-short > 0 and x1+short or w
		y2 = w-x1-short > 0 and h or w-x1
		line(x+x1,y+y1,x+x2-1,y+y2-1,c1)
	end
	for y1=2, h-1, 2 do
		x1 = 0
		y2 = h-y1-short > 0 and y1+short or h
		x2 = h-y1-short > 0 and w or h-y1
		line(x+x1,y+y1,x+x2-1,y+y2-1,c1)
	end
end
```

**Circle** - speed cca 0.22ms for radius 68, using bresenham algorithm. I modified code made by [Raidez](https://github.com/nesbox/tic.computer/wiki/code-examples-and-snippets#draw-circle-function)
```lua
function dcirc(x,y,r,c1,c2)
	local px, py, m = 0, r, 5-4*r
	if c2 then circ(x,y,r,c2) end
	local function drawLines ()
		if ((-py+x)%2==0 and (px+y)%2==0) or ((-py+x)%2~=0 and (px+y)%2~=0) then
			line((-py+x),(px+y),(-px+x),(py+y),c1)
			line((-py+x),(-px+y),(px+x),(py+y),c1)
		end
		if ((px+x)%2==0 and (-py+y)%2==0) or ((px+x)%2~=0 and (-py+y)%2~=0) then
			line((px+x),(-py+y),(py+x),(-px+y),c1)
			line((-px+x),(-py+y),(py+x),(px+y),c1)
		end
	end
	while (px<=py) do
		drawLines()
		if	m>0	then
			py=py-1
			m=m-8*py
			drawLines()
		end
		px=px+1
		m=m+8*px+4
	end
end
```

Propably not the best solution but only one which I made up. If you have any ideas how to improve it I will be glad :)

**Comment by Viza74** [comment]
@Crutiatix I took a shot at speeding up the rect routine. The idea was that maybe memcpy-ing lines is faster than drawing them, and it seems to be the case. Now, I must add that I'm veeeeery-veeeery tired now, and it is entirely probable that I'm screwed up something with the test or something, because it seems that this routine is faster than the built in normal filled rect, which sounds reeeealy stupid, so I don't know... o_O

Anyway, heres the code. Hold Z to draw with the new function (drect2)
If you uncomment line 14 and comment line 15, the comparsion is between the built-in rect, and drect2

```
local maxtime, mintime = 0,9999999999999
local maxtime2, mintime2 = 0,9999999999999

function TIC()
	cls()
	local test = 1
	if btn(4) then test=2 end

	local start=time()
	for i=1,1000 do
		if test==1 then
			--rect(0,0,240,136,9)
			drect(0,0,240,136,2,3)
		else
			drect2(0,0,240,136,2,3)
		end
	end
	local runtime = time()-start
	if test==1 then
		if runtime<mintime then mintime = runtime end
		if runtime>maxtime then maxtime = runtime end
		print("Test1",0,0,15)
	end
	if test==2 then
		if runtime<mintime2 then mintime2 = runtime end
		if runtime>maxtime2 then maxtime2 = runtime end
		print("Test2",0,0,15)
	end
	print("Test1 time: "..runtime.." max: "..maxtime.." min: "..mintime,0,125,15)
	print("Test2 time: "..runtime.." max: "..maxtime2.." min: "..mintime2,0,131,15)	
end

function drect2( x,y,w,h,c1,c2 )
	local rx,rw = x,w
	if (x+rw)%2~=0 then
		rw = rw-1
		dvline(x+w-1,y,y+h-2,c2,c1)
	end
	if rx%2~=0 then
		rx=rx+1
		rw=rw-1
		dvline(x,y,y+h-2,c1,c2)
		c1,c2=c2,c1
	end

	dhline(rx,y,rw-1,c1,c2)
	dhline(rx,y+1,rw-1,c2,c1)

	local w2 = (rw)/2+1
	local start = y*120+rx/2
	for i=0,h-1 do
		memcpy(start+i*120,start+(i%2)*120,w2)
	end
end

function dhline( x,y,length,c1,c2 )
	line(x,y,x+length,y,c2)

	for x1=x, x+length,2 do
		pix(x1,y,c1)
	end
end

function dvline( x,y,length,c1,c2 )
	line(x,y,x,y+length,c2)

	for y1=y, y+length-1,2 do
		pix(x,y1,c1)
	end
end

function drect (x,y,w,h,c1,c2)
	if c2 then rect(x,y,w,h,c2) end
	local short = w < h and w or h
	local x1, x2, y1, y2
	for x1=0, w-1, 2 do
		y1 = 0
		x2 = w-x1-short > 0 and x1+short or w
		y2 = w-x1-short > 0 and h or w-x1
		line(x+x1,y+y1,x+x2-1,y+y2-1,c1)
	end
	for y1=2, h-1, 2 do
		x1 = 0
		y2 = h-y1-short > 0 and y1+short or h
		x2 = h-y1-short > 0 and w or h-y1
		line(x+x1,y+y1,x+x2-1,y+y2-1,c1)
	end
end
```

**Comment by Crutiatix** [comment]
@Viza74 Nice! I done for sure a simpler speed test for your drect function and the result is that your function is really almost 3x faster than the built-in rect function 😆. So you didn't screwed up anything. 😄 I will start to use the memcpy function much more often.

**Comment by nesbox** [comment]
yes, I'm developing TIC without any optimization now, as is and there are a lot of ways to make drawing api run faster :)
btw, you can even use `memset` api to draw _dithered_ horizontal lines directly to the _screen buffer_, it will be the fastest way I think
``` lua
function hline(x,y,width,c1,c2)
 memset(x+240*y, c1|(c2<<4), width)
end
-- where c1 and c2 are colors
```

**Comment by Viza74** [comment]
I did a quick test, and it seems that memset is faster - or, more precisely the speed of memset and memcpy seems about the same, but with memset the routine can skip the "drawing by hand" of the first two horizontal lines, which is pretty slow because of the individual pix-ing.

**Comment by hseiken** [comment]
Nice!  Gonna have to pick apart these routines and exercise my headmuscle on them.

**Comment by Viza74** [comment]
New version: fixed some bugs, added some sanity checks, speed up with memset. Seems quite ok to use now. In fact it can be even used as a replacement for the built-in rect (passing the same color twice), since it is faster... but probably won't be faster for long, since it sounds more-or-less easy to optimize the engine-side rect for @nesbox 

```lua
function drect2( x,y,w,h,c1,c2 )
	if w<=0 or h<=0 or x>240 or x+w<=0 or y>136 or y+h<=0 then return end

	if x<0 then w=w+x x=0 end
	if y<0 then h=h+y y=0 end
	if x+w>240 then w=240-x end
	if y+h>136 then h=136-y end

	if w==1 then
		dvline(x,y,h,c1,c2)
		return
	end

	if h==1 then
		dhline(x,y,w,c1,c2)
		return
	end

	local rx,rw = x,w
	if x%2~=0 then
		rx=rx+1
		rw=rw-1
		dvline(x,y,h-1,c1,c2)
		c1,c2=c2,c1
	end

	if (x+w)%2~=0 then
		rw = rw-1
		dvline(x+w-1,y,h-1,c1,c2)
	end

	local dc1 = c1|(c2<<4)
	local dc2 = c2|(c1<<4)
	local w2 = (rw)/2
	local start = y*120+rx/2
	for i=0,h-1 do
		memset(start+i*120, (i%2==0) and dc1 or dc2,w2)
	end
end

function dhline( x,y,length,c1,c2 )
	line(x,y,x+length,y,c2)

	for x1=x, x+length,2 do
		pix(x1,y,c1)
	end
end

function dvline( x,y,length,c1,c2 )
	line(x,y,x,y+length,c2)

	for y1=y, y+length,2 do
		pix(x,y1,c1)
	end
end
```
Also, a little test to go with that:
```lua
function TIC()
	local xsize = 7.5
	local ysize = 7
	local offs = 0
	local voffs = 0

	cls()

	for o=0,18 do
		offs = math.floor(math.sin(o*ysize/10-time()/100)*(1+o*ysize/10))
		for i= 0, 15 do
			drect2(xsize*(i*2)+offs,  ysize*o, xsize,ysize, i,i)
			drect2(xsize*(i*2)+xsize+offs, ysize*o, xsize,ysize, i,i+1)
		end
	end

end
```

**Comment by hseiken** [comment]
Speaking of optimization, would this fix the draw problems I experience between maximized window and full screen on low-end windows machines?  PICO8 also has this problem.  I don't report it to either since multiple programs exhibit this behavior on my computer (which is quite frankly a bizarre machine so I won't get into that).  But there is a strange oddity that a maximized window has same display area + the windows decarations and desktop and runs 60fps but when going full screen, the frame rate drops to about 15 and everything runs slowly (i.e. it's not 'skipping' frames, it's drawing everything at a much slower rate).

Anyway, I wonder if that optimization was done if it would help on lower end systems like I have...it might be worth pursuing just to run on DIY hardware and such full screen.  ;)

**Comment by nesbox** [comment]
I think we can close it
